lib-js/spec-directory/virtual-directory.ts

```diff
@@ -0,0 +1,203 @@
+  }
+
+  async readFile(filename: string) {
+    return this.fileCache[filename]
+  }
+
+  async writeFile(filename: string, contents: string) {
+    if (this.subdirCache[filename]) {
+      throw new Error("Trying to write a subdirectory")
+    }
+    this.modified = true
+    this.fileCache[filename] = contents
+    if (!this.fileNames.includes(filename)) {
+      this.fileNames.push(filename)
+    }
+  }
+
+  async removeFile(filename: string) {
+    if (this.subdirCache[filename]) {
+      throw new Error("Trying to remove a subdirectory")
+    }
+    this.modified = true
+    delete this.fileCache[filename]
+    this.fileNames = this.fileNames.filter((f) => f !== filename)
+  }
+
+  // Subdir access
+
+  async listSubdirs() {
+    return this.subdirNames
+  }
+
+  async getSubdir(itemName: string) {
+    const subitem = this.subdirCache[itemName]
+    const options = await this.options()
+    if (!subitem) {
+      throw new Error(`Item does not exist: ${itemName}`)
+    }
+    return new VirtualDirectory(this.basePath, subitem, this.root, options)
+  }
+
+  // Iteration
+
+  // Write the files that are directly part of this directory
+  private async writeFilesToDisk() {
+    await fs.promises.mkdir(this.path, { recursive: true })
+    const files = await this.listFiles()
+    const writableFiles = files.filter((filename) => {
+      const { base, ext } = path.parse(filename)
+      if (base.startsWith("output")) return false
+      if (![".sass", ".scss", ".css"].includes(ext)) return false
+      return true
+    })
+    await Promise.all(
+      writableFiles.map(async (filename) => {
+        const filepath = path.resolve(this.path, filename)
+        await fs.promises.writeFile(filepath, await this.readFile(filename), {
+          encoding: "utf-8",
+        })
+      })
+    )
+  }
+
+  // To set up a virtual directory, write all files to disk
+  async setup() {
+    await this.writeFilesToDisk()
+    const subdirs = await this.subdirs()
+    await Promise.all(subdirs.map((subdir) => subdir.setup()))
+  }
+
+  // Return true if a this virtual directory has been modified
+  // (i.e. through interactive mode)
+  private async hasModifications(): Promise<boolean> {
+    const subdirs = await this.subdirs()
+    const subdirsNeedCleanup = await Promise.all(
+      subdirs.map(
+        (subdir) =>
+          subdir instanceof VirtualDirectory && subdir.hasModifications()
+      )
+    )
+    return this.modified || subdirsNeedCleanup.some((value) => value)
+  }
+
+  // Perform cleanup actions after opening this directory
+  async cleanup() {
+    // remove the physical directory
+    await fs.promises.rm(this.path, { recursive: true, force: true })
+    // if files were written to this directory, write to the root archive file
+    if (await this.hasModifications()) {
+      const hrx = await toHrx(this)
+      await fs.promises.writeFile(this.path + ".hrx", hrx, {
+        encoding: "utf-8",
+      })
+    }
```

Comment:
This seems like another unnecessary coupling. I understand the instinct to buffer all the modifications, but since interactive mode isn't performance-sensitive, I think you could simplify this a lot by just rewriting the entire HRX file every time the user performs a modification.